File: ros/src/system/autoware_health_checker/src/system_status_subscriber.cpp
/*
 * Copyright 2019 Autoware Foundation. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 * v1.0 Masaya Kataoka
 */

#include <autoware_health_checker/system_status_subscriber.h>

namespace autoware_health_checker {
SystemStatusSubscriber::SystemStatusSubscriber(ros::NodeHandle nh,
                                               ros::NodeHandle pnh) {
  nh_ = nh;
  pnh_ = pnh;
}

SystemStatusSubscriber::~SystemStatusSubscriber() {}

void SystemStatusSubscriber::enable() {
  ros::AsyncSpinner spinner(1);
  spinner.start();
  ros::Rate rate(1);
  status_sub_ = nh_.subscribe(
      "system_status", 10, &SystemStatusSubscriber::systemStatusCallback, this);
  while (ros::ok()) {
    rate.sleep();
  }
  spinner.stop();
  return;
}

void SystemStatusSubscriber::systemStatusCallback(
    const autoware_system_msgs::SystemStatus::ConstPtr msg) {
  for (auto function_itr = functions_.begin(); function_itr != functions_.end();
       function_itr++) {
    std::function<void(autoware_system_msgs::SystemStatus)> func =
        *function_itr;
    func(*msg);
  }
  return;
}

void SystemStatusSubscriber::addCallback(
    std::function<void(autoware_system_msgs::SystemStatus)> func) {
  functions_.push_back(func);
  return;
}
}

File: ros/src/system/autoware_health_checker/scripts/autoware_health_checker/system_status_subscriber.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright 2015-2019 Autoware Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import rospy
from autoware_system_msgs.msg import SystemStatus

class SystemStatusSubscriber:
    def __init__(self):
        self.__functions=[]

    def enable(self):
        rate = rospy.Rate(1)
        self.__status_sub = rospy.Subscriber("/system_status", SystemStatus, self.systemStatusCallback)
        #while not rospy.is_shutdown():
        #    rate.sleep()
        rospy.spin()

    def systemStatusCallback(self,msg):
        for function_itr in self.__functions:
            function_itr(msg)

    def addCallback(self,func):
        self.__functions.append(func)
